### PR TITLE
[ShopBundle] Logged customer after completing checkout can view order in his account

### PR DESCRIPTION
**Pull request by Tomanhez**
| Q               | A
| --------------- | -----
| Branch?         | 1.6
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| License         | MIT

Screenshot 2019-08-27 at 11 12 11

Screenshot 2019-08-27 at 15 53 51

### CODE REVIEW

File: src/Sylius/Bundle/ShopBundle/Resources/views/Order/thankYou.html.twig [hunk 1]
```diff
@@ -22,10 +22,13 @@
...
                 {% endif %}
                 <div class="ui hidden divider"></div>
-                <a href="{{ path('sylius_shop_order_show', {'tokenValue': order.tokenValue}) }}" id="sylius-show-order" class="ui large blue button">{{ 'sylius.ui.view_your_order_or_change_payment_method'|trans }}</a>
+                {% if order.customer.user is not null %}
+                    <a href="{{ path('sylius_shop_account_order_show', {'number': order.number}) }}" id="sylius-show-order-in-account" class="ui large blue button">{{ 'sylius.ui.view_your_order'|trans }}</a>
+                {% endif %}
             {% endif %}
 
             {% if order.customer.user is null %}
+                <a href="{{ path('sylius_shop_order_show', {'tokenValue': order.tokenValue}) }}" id="sylius-show-order" class="ui large blue button">{{ 'sylius.ui.view_your_order_or_change_payment_method'|trans }}</a>
```
Comment (Zales0123) [inline src/Sylius/Bundle/ShopBundle/Resources/views/Order/thankYou.html.twig]: Shouldn't there be only `sylius.ui.change_payment_method`? As viewing order is not possible for visitors?

File: src/Sylius/Behat/Context/Ui/Shop/AccountContext.php [hunk 2]
```diff
@@ -262,6 +262,14 @@ public function iBrowseMyOrders()
         $this->orderIndexPage->open();
     }
 
+    /**
+     * @When I try to open payment page
+     */
+    public function iTryToOpenPaymentPage()
```
Comment (Zales0123) [inline src/Sylius/Behat/Context/Ui/Shop/AccountContext.php]: Missing `: void`

File: src/Sylius/Behat/Context/Ui/Shop/AccountContext.php [hunk 3]
```diff
@@ -262,6 +262,14 @@ public function iBrowseMyOrders()
...
+    {
+        $this->orderShowPage->pay();
```
Comment (Zales0123) [inline src/Sylius/Behat/Context/Ui/Shop/AccountContext.php]: Is it paying or opening a payment page? For me, it's a little bit confusing what we try to do :/

**Comment by AdamKasp** [comment]
You should change your tests... in this task, you change redirect after checkout but tests check something else.

File: features/checkout/after_completing_checkout_being_able_to_go_on_order_in_my_accout.feature [hunk 4]
```diff
@@ -0,0 +1,20 @@
+@checkout
+Feature: After completing checkout being able to go on order in my account
```
Comment (CoderMaggie) [inline features/checkout/after_completing_checkout_being_able_to_go_on_order_in_my_accout.feature]:
```suggestion
Feature: Accessing order right after completing checkout
```

File: features/checkout/after_completing_checkout_being_able_to_go_on_order_in_my_accout.feature [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+    In order to managing my new order easily
```
Comment (CoderMaggie) [inline features/checkout/after_completing_checkout_being_able_to_go_on_order_in_my_accout.feature]:
```suggestion
    In order to check my order right after I have placed it
```

File: features/checkout/after_completing_checkout_being_able_to_go_on_order_in_my_accout.feature [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+    As a Customer
+    I want to being able to go on order in my account after completing checkout
```
Comment (CoderMaggie) [inline features/checkout/after_completing_checkout_being_able_to_go_on_order_in_my_accout.feature]:
```suggestion
    I want to access this order from the thank you page
```

File: features/checkout/after_completing_checkout_being_able_to_go_on_order_in_my_accout.feature [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+    Background:
+        Given the store operates on a single channel in "United States"
+        And there is a user "john@example.com" identified by "password123"
+        And the store allows paying "Cash on delivery"
+        And the store has a product "PHP T-Shirt" priced at "$19.99"
+        And the store ships everywhere for free
+        And I am logged in as "john@example.com"
+
+    @ui
+    Scenario: Being able to go on order in my account after completing checkout
```
Comment (CoderMaggie) [inline features/checkout/after_completing_checkout_being_able_to_go_on_order_in_my_accout.feature]:
```suggestion
    Scenario: Being able to access my order right after completing checkout
```

File: features/checkout/after_completing_checkout_being_able_to_go_on_order_in_my_accout.feature [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+        Given I added product "PHP T-Shirt" to the cart
+        And I have proceeded selecting "Cash on delivery" payment method
+        When I have confirmed order
+        Then I should be able to go to order details in my account
```
Comment (CoderMaggie) [inline features/checkout/after_completing_checkout_being_able_to_go_on_order_in_my_accout.feature]:
```suggestion
        Then I should be able to access this order's details
```

File: features/checkout/paying_for_order/changing_payment_method_after_order_confirmation_item_quantity.feature [hunk 9]
```diff
@@ -6,16 +6,15 @@ Feature: Having good number of items in changing payment method page
 
     Background:
         Given the store operates on a single channel in "United States"
-        And there is a user "john@example.com" identified by "password123"
         And the store allows paying "Offline"
         And the store allows paying "Cash on delivery"
         And the store has a product "PHP T-Shirt" priced at "$19.99"
         And the store ships everywhere for free
-        And I am logged in as "john@example.com"
```
Comment (GSadee) [inline features/checkout/paying_for_order/changing_payment_method_after_order_confirmation_item_quantity.feature]: So you should change to `As a Guest/Visitor`

File: src/Sylius/Behat/Page/Shop/Order/ThankYouPage.php [hunk 10]
```diff
@@ -25,6 +25,11 @@ public function goToOrderDetails()
         $this->getElement('order_details')->click();
     }
 
+    public function goToOrderDetailsInMyAccount(): void
```
Comment (GSadee) [inline src/Sylius/Behat/Page/Shop/Order/ThankYouPage.php]:
```suggestion
    public function goToOrderDetailsInAccount(): void
```

File: src/Sylius/Bundle/ShopBundle/Resources/views/Order/thankYou.html.twig [hunk 11]
```diff
@@ -22,10 +22,13 @@
...
                 {% endif %}
                 <div class="ui hidden divider"></div>
-                <a href="{{ path('sylius_shop_order_show', {'tokenValue': order.tokenValue}) }}" id="sylius-show-order" class="ui large blue button">{{ 'sylius.ui.view_your_order_or_change_payment_method'|trans }}</a>
+                {% if order.customer.user is not null %}
+                    <a href="{{ path('sylius_shop_account_order_show', {'number': order.number}) }}" id="sylius-show-order-in-account" class="ui large blue button">{{ 'sylius.ui.view_your_order'|trans }}</a>
+                {% endif %}
             {% endif %}
 
             {% if order.customer.user is null %}
+                <a href="{{ path('sylius_shop_order_show', {'tokenValue': order.tokenValue}) }}" id="sylius-show-order" class="ui large blue button">{{ 'sylius.ui.change_payment_method'|trans }}</a>
```
Comment (GSadee) [inline src/Sylius/Bundle/ShopBundle/Resources/views/Order/thankYou.html.twig]: Should it be moved from `{% if lastPayment != false %}`?

File: src/Sylius/Bundle/UiBundle/Resources/translations/messages.en.yml [hunk 12]
```diff
@@ -902,6 +902,7 @@ sylius:
         view_and_edit_cart: 'View and edit cart'
         view_more: 'View more'
         view_product_details: 'View %product% details'
+        view_your_order: 'View order'
```
Comment (GSadee) [inline src/Sylius/Bundle/UiBundle/Resources/translations/messages.en.yml]:
```suggestion
        view_order: 'View order'
```

File: src/Sylius/Behat/Page/Shop/Order/ThankYouPage.php [hunk 13]
```diff
@@ -74,13 +79,11 @@ public function getRouteName(): string
...
     }
 
-    /**
-     * {@inheritdoc}
-     */
     protected function getDefinedElements(): array
     {
         return array_merge(parent::getDefinedElements(), [
             'order_details' => '#sylius-show-order',
+            'order_details_in_my_account' => '#sylius-show-order-in-account',
```
Comment (GSadee) [inline src/Sylius/Behat/Page/Shop/Order/ThankYouPage.php]:
```suggestion
            'order_details_in_account' => '#sylius-show-order-in-account',
```

File: src/Sylius/Bundle/UiBundle/Resources/translations/messages.en.yml [hunk 14]
```diff
@@ -902,6 +902,7 @@ sylius:
         view_and_edit_cart: 'View and edit cart'
         view_more: 'View more'
         view_product_details: 'View %product% details'
+        view_order: 'View order'
```
Comment (GSadee) [inline src/Sylius/Bundle/UiBundle/Resources/translations/messages.en.yml]: Wrong order of keys

File: features/checkout/paying_for_order/accessing_order_right_after_completing_checkout.feature [hunk 15]
```diff
@@ -1,21 +1,20 @@
...
         And the store ships everywhere for free
-        And I am logged in as "john@example.com"
 
     @ui
     Scenario: Seeing correct quantity on payment retry page
         Given I have added 2 products "PHP T-Shirt" to the cart
+        And I complete addressing step with email "john@example.com" and "United States" based shipping address
         And I have proceeded selecting "Cash on delivery" payment method
         And I have confirmed order
         When I go to order details
```
Comment (lchrusciel) [inline features/checkout/paying_for_order/accessing_order_right_after_completing_checkout.feature]: How can we go to order details, if we are not a registered customer?

File: src/Sylius/Behat/Context/Ui/Shop/Checkout/CheckoutThankYouContext.php [hunk 16]
```diff
@@ -44,6 +44,14 @@ public function iProceedToTheRegistration(): void
         $this->thankYouPage->createAccount();
     }
 
+    /**
+     * @Then I should be able to access this order's details
+     */
+    public function iShouldBeAbleToAccessThisOrderSDetails(): void
+    {
+        $this->thankYouPage->goToOrderDetailsInAccount();
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Ui/Shop/Checkout/CheckoutThankYouContext.php]: We need some assertion here. This step is not testing anything.

File: src/Sylius/Bundle/ShopBundle/Resources/views/Order/thankYou.html.twig [hunk 17]
```diff
@@ -22,10 +22,15 @@
                     </div>
                 {% endif %}
                 <div class="ui hidden divider"></div>
-                <a href="{{ path('sylius_shop_order_show', {'tokenValue': order.tokenValue}) }}" id="sylius-show-order" class="ui large blue button">{{ 'sylius.ui.view_your_order_or_change_payment_method'|trans }}</a>
+                {% if order.customer.user is not null %}
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ShopBundle/Resources/views/Order/thankYou.html.twig]: This section should be outside of `{% if lastPayment != false %}`

Comment (Tomanhez) [reply src/Sylius/Bundle/ShopBundle/Resources/views/Order/thankYou.html.twig]: I did that but @GSadee suggested changes

File: features/checkout/paying_for_order/accessing_order_right_after_completing_checkout.feature (hunk 15 above)
Comment (Tomanhez) [reply features/checkout/paying_for_order/accessing_order_right_after_completing_checkout.feature]: I need pass address steep

File: src/Sylius/Bundle/ShopBundle/Resources/views/Order/thankYou.html.twig [hunk 18]
```diff
@@ -22,10 +22,13 @@
...
```
Comment (pamil) [inline src/Sylius/Bundle/ShopBundle/Resources/views/Order/thankYou.html.twig]: Could we move that `if` outside the `if` with `lastPayment != false`? We can group it together with the other one below and use `if` and `else`.

File: src/Sylius/Behat/Context/Ui/Shop/Checkout/CheckoutThankYouContext.php [hunk 19]
```diff
@@ -44,6 +49,18 @@ public function iProceedToTheRegistration(): void
         $this->thankYouPage->createAccount();
     }
 
+    /**
+     * @Then /^I should be able to access this order's details$/
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Ui/Shop/Checkout/CheckoutThankYouContext.php]:
```suggestion
     * @Then I should be able to access this order's details
```

File: src/Sylius/Behat/Page/Shop/Order/ThankYouPage.php [hunk 20]
```diff
@@ -74,13 +79,11 @@ public function getRouteName(): string
...
 
-    /**
-     * {@inheritdoc}
-     */
     protected function getDefinedElements(): array
     {
         return array_merge(parent::getDefinedElements(), [
-            'order_details' => '#sylius-show-order',
+            'payment_method_page' => '#payment-method-page',
+            'order_details_in_account' => '#sylius-show-order-in-account',
```
Comment (GSadee) [inline src/Sylius/Behat/Page/Shop/Order/ThankYouPage.php]: Wrong order of keys

File: src/Sylius/Behat/Context/Ui/Shop/Checkout/CheckoutThankYouContext.php [hunk 21]
```diff
@@ -44,6 +49,18 @@ public function iProceedToTheRegistration(): void
...
+     */
+    public function iShouldBeAbleToAccessThisOrderDetails(): void
+    {
+        $this->thankYouPage->goToOrderDetailsInAccount();
+
+        $number = $this->orderShowPage->getNumber();
+
+        Assert::true($number !== null ? true : false);
```
Comment (Zales0123) [inline src/Sylius/Behat/Context/Ui/Shop/Checkout/CheckoutThankYouContext.php]: 😢

File: src/Sylius/Behat/Context/Ui/Shop/Checkout/CheckoutThankYouContext.php [hunk 22]
```diff
@@ -44,6 +57,18 @@ public function iProceedToTheRegistration(): void
...
+    /**
+     * @Then I should be able to access this order's details
+     */
+    public function iShouldBeAbleToAccessThisOrderDetails(): void
+    {
+        $this->thankYouPage->goToOrderDetailsInAccount();
+
+        $number = $this->orderShowPage->getNumber();
+
+        Assert::true($number === $this->orderRepository->findLatest(1)[0]->getNumber());
```
Comment (Zales0123) [inline src/Sylius/Behat/Context/Ui/Shop/Checkout/CheckoutThankYouContext.php]: `Assert::same($this->orderRepository->findLatest(1)[0]->getNumber(), $number);`

**Comment by pamil** [comment]
Thank you, Tomasz! :tada:

**Pull request merged**